### ISSUE TITLE
Crash when waking up laptop from suspend while using external monitor.

### ISSUE THREAD

**Issue by rszyma** [hyprwm/Hyprland#3076]
Started getting these crashes after getting an external monitor (I disabled laptop screen if that matters). I don't know exactly when the crash happens, but when I wake from suspend I see crash in tty, so it's probably some time after my laptop suspends or right before waking it up. 
[hyprlandCrashReport525.txt](https://github.com/hyprwm/Hyprland/files/12250696/hyprlandCrashReport525.txt)

_Originally posted by @rszyma in https://github.com/hyprwm/Hyprland/issues/2745#issuecomment-1663883384_

____

I'm reposting this in a new issue, because crash reports I get are consistent with their content, while the 2 other crash reports in #2745 are entirely different. Also steps to reproduce slightly different in my case.

Some other crash reports (notice that they all have similar log tail):
[hyprlandCrashReport65117.txt](https://github.com/hyprwm/Hyprland/files/12441496/hyprlandCrashReport65117.txt)
[hyprlandCrashReport741.txt](https://github.com/hyprwm/Hyprland/files/12441504/hyprlandCrashReport741.txt)
[hyprlandCrashReport744.txt](https://github.com/hyprwm/Hyprland/files/12441512/hyprlandCrashReport744.txt)
[hyprlandCrashReport574.txt](https://github.com/hyprwm/Hyprland/files/12441514/hyprlandCrashReport574.txt)

___

Some additional notes:
- The crash sometimes happens, sometimes not. I'd say that it's about 70% of times when leaving my laptop suspended overnight.
- I've tried `systemctl suspend` and then waking up immediately, but then the crash don't show up. From my testing it seems that the crash can happen only after my laptop is suspended when it's asleep for at least ~1 minute.
- The monitor I use is [this](https://www.gigabyte.com/Monitor/G24F-2/sp#sp). I haven't tried reproducing with other monitors.
- The crash no longer happens when I enable laptop monitor.

___

Monitors config I use:
```
monitor=eDP-1,disable
monitor=HDMI-A-1,1920x1080@180,0x0,1
```

**Comment by mauricekraus** [comment]
Same thing can be observed on my setup. I have two external monitors and additionally run kanshi to auto enable/disable my monitors.

I also added exemplary two reports

[hyprlandCrashReport152128.txt](https://github.com/hyprwm/Hyprland/files/12452350/hyprlandCrashReport152128.txt)
[hyprlandCrashReport181356.txt](https://github.com/hyprwm/Hyprland/files/12452351/hyprlandCrashReport181356.txt)

--------------

Some additional notes:

> The crash sometimes happens, sometimes not. I'd say that it's about 70% of times when leaving my laptop suspended overnight. 

Same behavior

> I've tried systemctl suspend and then waking up immediately, but then the crash don't show up. From my testing it seems that the crash can happen only after my laptop is suspended when it's asleep for at least ~1 minute.

Same behavior

> The monitor I use is [this](https://www.gigabyte.com/Monitor/G24F-2/sp#sp). I haven't tried reproducing with other monitors. 

I use one LG Ultrawide 32 Inch (via usbc) and one vertical LG 28 Inch (slightly different refresh rates) 

> The crash no longer happens when I enable laptop monitor.

Not tested

Kanshi config
```
profile home {
	output "LG Electronics LG ULTRAWIDE 0x0004A32C" mode 3440x1440@59.973Hz position 0,0
	output "LG Electronics LG ULTRAGEAR 110MALFNX072" mode 2560x1440@59.951Hz 
  output "eDP-1" disable
}


profile undocked {
    output "eDP-1" enable position 0,0
}

```

And in addition i use this shell script to enable my laptop monitor on disconnect (if I do that via kanshi hyprland crashes as well)

```bash
#!/bin/sh

handle() {
  if [ "$1" = "monitorremoved>>eDP-1" ]; then
    return
  fi

  connected_mons=$(hyprctl monitors | grep Monitor | wc -l)
  case $1 in
    monitoradded*) echo "Monitor added $1, now $connected_mons" ;;
    monitorremoved*)
      echo "Monitor removed $1, now $connected_mons"
      hyprctl keyword monitor eDP-1,$mon2info
      ;;
  esac
}

socat -U - UNIX-CONNECT:/tmp/hypr/$HYPRLAND_INSTANCE_SIGNATURE/.socket2.sock | while read -r line; do handle "$line"; done

```

In my hyprland config the monitors are setup in the following way:

```
monitor=,preferred,auto,auto
monitor=desc:LG Electronics LG ULTRAWIDE 0x0004A32C, 3440x1440@59.973, 0x0, 1.0 # ULTRAWIDE HOME
monitor=desc:LG Electronics LG ULTRAGEAR 110MALFNX072, 2560x1440@59.951, 3440x-550, 1.0, transform,3 #VERTICAL HOME

env = mon2info,3840x2160@60.007000,0x0,2.5 # needed for enabling
```

Maybe this further helps to track down the issue.

**Comment by coffebar** [comment]
[crash1.txt](https://github.com/hyprwm/Hyprland/files/12505820/crash.txt)
[crash2.txt](https://github.com/hyprwm/Hyprland/files/12505827/crash2.txt)

Starting from 2023-09-02 I see a crash when the external monitor powered on (after power off or was switched to another input).

It's not reproducible but occurs only when the laptop adds an external monitor, that is already physically connected, but for some reason was not working with this machine (power outage or switch signal to another machine).

**Comment by CoolnsX** [comment]
same crashes on my laptop, when connected to external monitor, I have script that detects the monitor and move the workspaces accordingly, whenever my monitor disconnects and re-connects due to power cut, it crashes my hyprland

**Comment by vaxerski** [comment]
can someone test on 0.29.1?

**Comment by rszyma** [comment]
I just tested on 4eecb8bffc829ab2222b4acea1942ef8ee7b8904 and there's a different problem now. 

If I have running electron-based app (like vivaldi, vscode or obsidian) it and suspend and wake up (or turn off and turn back on the monitor) Hyprland crashes 
[hyprlandCrashReport16487.txt](https://github.com/hyprwm/Hyprland/files/12515967/hyprlandCrashReport16487.txt)

Otherwise (if no electron-based program is running), suspending and wake up (or turning off and turning back on the monitor) closes all opened apps (but no crash)

**Comment by vaxerski** [comment]
this crash is from some invalid hyprctl call with poor timing

**Comment by rszyma** [comment]
ok, but why are all the programs closing? This wasn't the case before

**Comment by vaxerski** [comment]
![image](https://github.com/hyprwm/Hyprland/assets/43317083/23ff8572-2347-48ec-9e73-7b9b8107131c)

launch one from a shell and log output to a file and check what went wrong

**Comment by rszyma** [comment]
Actually looks like the programs don't get closed. It's just that the active workspace is somehow changed to the next free workspace. E.g. if I have things opened on 1st and 2nd workspaces, Hyprland will switch me to 3rd workspace. I guess I could open a separate issue just for that, but I don't care, it's not as annoying for me.

**Comment by rszyma** [comment]
> this crash is from some hyprctl call with poor timing

would me enabling some debug flag show which call is that? Or anything else I could do to help fix it?

**Comment by vaxerski** [comment]
I don't know, honestly. What are the calls you have in your config / executed?
This looks like `hyprctl activeworkspace` on an inactive session

**Comment by rszyma** [comment]
I don't have any `hyprctl` calls in my config. 

I did some more testing and it looks like it's not just electron apps, but more generally, apps run via xwayland. 

To confirm this, I've tested with [xeyes](https://gitlab.freedesktop.org/xorg/app/xeyes) left running and turned monitor off and on. One time I got Hyprland crash and another time Hyprland didn't crash but the `xeyes` exited and left an error message in terminal: 
```
X connection to :0 broken (explicit kill or server shutdown).
```

**Comment by mauricekraus** [comment]
After testing your patch, indeed, Hyprland no longer crashes. However, I can observe the same behavior as rszyma. All XWayland Apps and Electron Apps exit, such as Zoom, Edge (I run it on Xwayland), VSCode, or Mattermost. Hence, it's not only Electron-based Apps. I've also noticed that my terminals with Neovim running are getting closed. Foot should neither run on XW nor Electron. Furthermore, Waylock (-effects) crashes 100% of the time and leaves the Desktop unlocked. Could you please tell me what I can do to further track down the issue?

**Comment by CoolnsX** [comment]
so after updating to 0.29.1, the crashing problem solved, but now whenever my monitor disconnects/re-connects it closes all the application on all workspaces

**Comment by alvaro17f** [comment]
@CoolnsX same here

**Comment by vaxerski** [comment]
can yall test -git from today

**Comment by rszyma** [comment]
> can yall test -git from today

yep, works

**Comment by mauricekraus** [comment]
Works great, from my side as well! Thank you @vaxerski. However, Swaylock is still crashing from time to time, but i don't think this is a hyprland issue.
